// brace/mode/dolittle.js
ace.define(
  "ace/mode/dolittle_highlight_rules",
  [
    "require",
    "exports",
    "module",
    "ace/lib/oop",
    "ace/mode/text_highlight_rules"
  ],
  function(require, exports, module) {
    "use strict";

    var oop = require("../lib/oop");
    var TextHighlightRules = require("./text_highlight_rules")
      .TextHighlightRules;

    var DolittleHighlightRules = function() {
      var object =
        "タイマー|配列|画面|背景" +
        "|色|光|絵具" +
        "|黒|赤|緑|青|黄色|黄|紫|水色|白" +
        "|タートル" +
        "|ラベル|ボタン|フィールド|リスト" +
        "|自分|システム|ルート|シリアルポート" +
        "|マイクロビット|MBit" +
        "|Timer|Array|Panel|Screen" +
        "|paint|light|pigment|Bluetooth" +
        "|Black|Red|Green|Blue|Yellow|Magenta|Cyan|White" +
        "|turtle|List|self" +
        "|Label|Button|Field|system|root" +
        "|Serialport";

      var method =
        "作る|繰り返す|なら|そうでなければ|の間|実行" +
        "|歩く|戻る|移動する|位置|中心に戻る|円|角形|閉じる|塗る" +
        "|右回り|左回り|向き" +
        "|ペンあり|ペンなし|線の色|線の太さ" +
        "|図形を作る|拡大する|消える|現れる|手前に表示|変身する" +
        "|方眼紙" +
        "|明るくする|暗くする" +
        "|使う|接続する|書く|動作|読む|接続|次の行|クリア" +
        "|向き？|横の位置？|縦の位置？|反対" +
        "|動作|衝突|跳ね返る|開く|ポート選択|自動選択" +
        "|間隔|回数|時間|次に実行|最後に実行" +
        "|横の位置|縦の位置|閉じる|部分|長さ|文字コード|コード文字" +
        "|受信|送信|次に実行" +
        "|create|repeat|then|else|while|execute" +
        "|forward|backward|moveBy|moveTo|movetocenter|circle|polygon|closepath" +
        "|rightturn|leftturn|direction" +
        "|penUp|penDown|lineColor|lineWidth" +
        "|makeFigure|scale|hide|show|movefront|setshape" +
        "|showgrid|write|read" +
        "|use" +
        "|brighter|darker" +
        "|substr|contain|indexof|nextLine|action|charcode|next_execute";

      var keywordboolean = "true|false|はい|いいえ";

      var keywordMapper = this.createKeywordMapper(
        {
          list: object,
          variable: method,
          constant: keywordboolean
        },
        "identifier",
        false
      );

      this.$rules = {
        start: [
          {
            defaultToken: "text"
          },
          {
            token: "list",
            regex: "ステータス|振動計|方角"
          },
          {
            token: "variable",
            regex: "SerialportSelect|シリアルポート選択|動作設定"
          },
          {
            token: "list",
            regex: object
          },
          {
            token: "variable",
            regex: method
          },
          {
            token: "constant.language.boolean",
            regex: keywordboolean
          },
          {
            token: "comment",
            regex: /\/\/.*$/
          },
          {
            token: "comment",
            regex: /\/\*/,
            next: "comment"
          },
          {
            token: "string",
            regex: '".*?"'
          },
          {
            token: "string",
            regex: "”.*?”"
          },
          {
            token: "string",
            regex: "'.*?'"
          },
          {
            token: "constant.numeric",
            regex: "[+-]?\\d+(?:(?:\\.\\d*)?(?:[eE][+-]?\\d+)?)?\\b"
          },
          {
            token: "constant.numeric",
            regex: "0[xX][0-9a-fA-F]+(L|l|UL|ul|u|U|F|f|ll|LL|ull|ULL)?\\b"
          },
          {
            token: "constant.numeric",
            regex: "[０-９]"
          },
          {
            token: "keyword.operator",
            regex:
              "\\+|\\-|\\/|\\/\\/|%|<@>|@>|<@|&|\\^|~|<|>|<=|=>|==|!=|<>|=" +
              "|!|！|\\.|．|。|：|:|＝"
          },
          {
            token: "paren.lparen",
            regex: "[\\(]"
          },
          {
            token: "paren.rparen",
            regex: "[\\)]"
          },
          {
            token: "text",
            regex: "\\s+"
          }
        ],
        comment: [
          {
            defaultToken: "comment"
          },
          {
            token: "comment",
            regex: /\*\//,
            next: "start"
          }
        ]
      };
    };

    oop.inherits(DolittleHighlightRules, TextHighlightRules);

    exports.DolittleHighlightRules = DolittleHighlightRules;
  }
);

ace.define(
  "ace/mode/dolittle",
  [
    "require",
    "exports",
    "module",
    "ace/lib/oop",
    "ace/mode/text",
    "ace/mode/dolittle_highlight_rules"
  ],
  function(require, exports, module) {
    "use strict";

    var oop = require("../lib/oop");
    var TextMode = require("./text").Mode;
    var DolittleHighlightRules = require("./dolittle_highlight_rules")
      .DolittleHighlightRules;

    var Mode = function() {
      this.HighlightRules = DolittleHighlightRules;
      this.$behaviour = this.$defaultBehaviour;
    };
    oop.inherits(Mode, TextMode);

    (function() {
      this.lineCommentStart = "//";
      this.blockComment = {
        start: "/*",
        end: "*/"
      };
      this.$id = "ace/mode/dolittle";
    }.call(Mode.prototype));

    exports.Mode = Mode;
  }
);
